src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -289,112 +506,49 @@ private void AnalyzeMethod(Action<Diagnostic> reportDiagnostic, IMethodSymbol me
-                        }
-                    }
-                    // For non-C# languages, we'll do a quick check to catch simple cases
-                    // since IsUnmanagedType only works in languages that support unmanaged types
-                    // and non-C# languages that might not support is (such as VB) aren't a big focus of the attribute
-                    // this analyzer validates.
-                    else if (type.IsReferenceType || type.GetMembers().Any(m => m is IFieldSymbol { IsStatic: false, Type.IsReferenceType: true }))
-                    {
-                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabledManagedParameterOrReturnTypes));
-                    }
-
-                    if (type.IsValueType && TypeIsAutoLayoutOrContainsAutoLayout(type))
-                    {
-                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabledAutoLayoutTypes));
-                    }
+                    reportDiagnostic(paramLocation.CreateDiagnostic(descriptorOverride ?? FeatureUnsupportedWhenRuntimeMarshallingDisabledByRefParameters));
                 }
+                AnalyzeSignatureType(paramLocation, param.Type);
             }
 
-            private bool TypeIsAutoLayoutOrContainsAutoLayout(ITypeSymbol type)
+            void AnalyzeSignatureType(ImmutableArray<Location> locations, ITypeSymbol type)
             {
-                return TypeIsAutoLayoutOrContainsAutoLayout(type, ImmutableHashSet<ITypeSymbol>.Empty.WithComparer(SymbolEqualityComparer.Default));
-                bool TypeIsAutoLayoutOrContainsAutoLayout(ITypeSymbol type, ImmutableHashSet<ITypeSymbol> seenTypes)
+                if (type.SpecialType == SpecialType.System_Void)
                 {
-                    Debug.Assert(type.IsValueType);
-
-                    if (_isAutoLayoutOrContainsAutoLayoutCache.TryGetValue(type, out bool isAutoLayoutOrContainsAutoLayout))
-                    {
-                        return isAutoLayoutOrContainsAutoLayout;
-                    }
-
-                    if (seenTypes.Contains(type.OriginalDefinition))
-                    {
-                        // If we have a recursive type, we are in one of two scenarios.
-                        // 1. We're analyzing CoreLib and see the struct definition of a primitive type.
-                        // In all of these cases, the type does not have auto layout.
-                        // 2. We found a recursive type definition.
-                        // Recursive type definitions are invalid and Roslyn will emit another error diagnostic anyway,
-                        // so we don't care here.
-                        _isAutoLayoutOrContainsAutoLayoutCache.TryAdd(type, false);
-                        return false;
-                    }
-
-                    if (_structLayoutAttribute is not null)
```

Comment:
Ah, I see this was existing code just moved around. Feel free to ignore my comments above if you don't want to change existing code.